feat: add prefer-logical-properties rule

azat-io opened the pull request:
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Add new rule.

#### What changes did you make? (Give an overview)

Added new rule for usage logical CSS properties instead of phisical.

#### Related Issues

#13

#### Is there anything you'd like reviewers to focus on?

I don't think so.

nzakas commented:
Thanks for getting this started. There were some edge cases that were discussed in the original issue:

https://github.com/eslint/css/issues/13#issuecomment-2488917780
https://github.com/eslint/css/issues/13#issuecomment-2489258859

Can you describe how you're handling those?

nzakas commented:
I think we need a way for people to opt-out on a case-by-case basis. At a minimum, I think we'd need to be able to do this:

```js
{
    "prefer-logical-properties": ["error", {
        allowedProperties: ["padding-left"],   // don't warn on these
        allowedUnits: ["xyz"]  // don't warn on these
    }
}
```

nzakas commented on src/rules/prefer-logical-properties.js:
```diff
@@ -0,0 +1,175 @@
+// Helpers
+//-----------------------------------------------------------------------------
+
+const propertiesReplacements = {
```
Please move all of these lookup tables into maps for better performance.

azat-io replied:
Done ✅

azat-io commented:
I've added options `allowedUnits` and `allowedProperties` ✅.

nzakas commented on src/rules/prefer-logical-properties.js:
```diff
@@ -0,0 +1,232 @@
+	meta: {
+		type: /** @type {const} */ ("problem"),
+
+		fixable: "code",
```
To fix the type error.
```suggestion
		fixable: /** @type {const} */ ("code"),
```

azat-io replied:
Fixed ✅

azat-io commented:
Renamed the options and added their descriptions to the documentation.

nzakas commented on docs/rules/prefer-logical-properties.md:
Just to finish up the docs, we typically include two additional sections:

1. When Not to Use It - gives guidance on when it's okay to not use the rule
2. Prior Art - links to other tools that do the same thing

Here's an example:
https://github.com/eslint/css/blob/main/docs/rules/no-invalid-properties.md

For prior art, we have:

* https://github.com/csstools/stylelint-use-logical
* https://www.npmjs.com/package/stylelint-plugin-logical-css

azat-io replied:
Added ✅

nzakas commented:
This is looking good. I just realized that it will also error for `@supports` rules, such as:

```css
@supports (text-align: left) {
    /* rules */
}
```

I'm not sure if that's the correct behavior. What do you think?

jeddy3 commented:
Looking good! Two questions...

Do we want the rule to also flag the [`width`](https://drafts.csswg.org/css-conditional-5/#width) and [`height`](https://drafts.csswg.org/css-conditional-5/#descdef-container-height) [size features](https://drafts.csswg.org/css-conditional-5/#size-container) (useful for alignment with the [container type](https://drafts.csswg.org/css-conditional-5/#container-type))?, i.e. allow this:

```css
main {
  container-type: inline-size;

  & h2 {
    @container (inline-size > 40em) {
      /* .. */
    }
  }
```

While disallowing this:


```css
main {
  container-type: inline-size;

  & h2 {
    @container (width > 40em) {
      /* .. */
    }
  }
```

And do we want to do anything special with `overflow-x` and `overflow-y` as their [logical counterparts are only supported by Firefox](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-inline#browser_compatibility)?

nzakas commented:
> And do we want to do anything special with `overflow-x` and `overflow-y` as their [logical counterparts are only supported by Firefox](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-inline#browser_compatibility)?

I think we can skip these for now. If/when other browsers implement the logical equivalent we can revisit.

> Do we want the rule to also flag the [`width`](https://drafts.csswg.org/css-conditional-5/#width) and [`height`](https://drafts.csswg.org/css-conditional-5/#descdef-container-height) [size features](https://drafts.csswg.org/css-conditional-5/#size-container) (useful for alignment with the [container type](https://drafts.csswg.org/css-conditional-5/#container-type))?,

These aren't properties, so I'm not sure it makes sense to bundle this check in this rule.

azat-io commented:
> This is looking good. I just realized that it will also error for @supports rules

You are right. We should not use linting in `@supports` directives. Stylelint's rules don't do that either.

I've updated the PR and added new tests.

<img width="299" alt="image" src="https://github.com/user-attachments/assets/64fe5716-dea4-4e89-a3a5-39f58004a6da" />

nzakas reviewed (approved):
LGTM. Thanks for all of your work on this.

(The CI failure is unrelated to this change, so we're not going to worry about that.)
